source/sliding_puzzle.py

```diff
@@ -124,7 +139,88 @@ def _join(self, row: List[PuzzlePiece]) -> List[str]:
     def draw(self) -> str:
         """Draw the puzzle in its current state."""
         output = []
-        for row in self.pieces:
+        for row in self.rows:
             output.extend(self._join(row))
 
         return '\n'.join(output)
+
+    def shuffle(self) -> None:
+        """Shuffle the puzzle by randomizing the order of the pieces"""
+        for row in self.rows:
+            random.shuffle(row)
+        puzzle_heigth, puzzle_width = len(self.rows), len(self.rows[0])
+        # empty one of the rows
+        x = random.choice(range(puzzle_heigth))
+        y = random.choice(range(puzzle_width))
+        self.rows[x][y].clear()
+
+    def _get_empty_piece_position(self) -> PiecePosition:
+        x, y, index = [
+            (x, y, piece.index)
+            for x, row in enumerate(self.rows)
+            for y, piece in enumerate(row)
+            if piece._data == []
+        ][0]
+        return PiecePosition(x=x, y=y, index=index)
+
+    def move_down(
+        self,
+    ) -> None:
+        """Move the piece above the empty box down"""
+        empty_pos = self._get_empty_piece_position()
+        # return if empty piece is on the first row
+        top_row = self.rows[0]
+        if empty_pos.index in [piece.index for piece in top_row]:
+            return
+        # swap the empty piece with the target piece
+        self.rows[empty_pos.x][empty_pos.y], self.rows[empty_pos.x - 1][empty_pos.y] = (
+            self.rows[empty_pos.x - 1][empty_pos.y],
+            self.rows[empty_pos.x][empty_pos.y],
+        )
```

Comment:
This tuple assignment actually ends up taking up more space than just doing
```py
x = 1
y = 2
```

Reply:
Since commenting this I did realize why we have to do it this way. Otherwise we'd need a 3rd temp variable. Perhaps move this code into a method that takes 2 tuples (or 4 integers) and switches the position. That way we can do: `self.move((x, y), (x-1, y))`

Reply:
that's a good idea